File: Classes/CSVParser.m
// CSVParser.m
//
// Kraftstoff


#import "CSVParser.h"


@implementation CSVParser
{
	NSString *csvString;
	NSString *separator;
	NSScanner *scanner;

	NSMutableArray *fieldNames;
	NSCharacterSet *endTextCharacterSet;
}


+ (NSString *)simplifyCSVHeaderName:(NSString *)header
{
    return [[header stringByReplacingOccurrencesOfString:@"[? :_\\-]+"
                                              withString:@""
                                                 options:NSRegularExpressionSearch
                                                   range:NSMakeRange (0, [header length])] uppercaseString];
}


- (instancetype)initWithString:(NSString *)inputCSVString
{
    if ((self = [super init])) {

        // Convert DOS and legacy Mac line endings to Unix
        csvString = [inputCSVString stringByReplacingOccurrencesOfString:@"\r\n?"
                                                              withString:@"\n"
                                                                 options:NSRegularExpressionSearch
                                                                   range:NSMakeRange (0, [inputCSVString length])];

        scanner = [[NSScanner alloc] initWithString:csvString];
        [scanner setCharactersToBeSkipped:nil];
    }

    return self;
}


- (void)revertToBeginning
{
    [scanner setScanLocation:0];
}


- (void)setSeparator:(NSString *)separatorString
{
    if (! [separator isEqualToString:separatorString]) {
        separator = separatorString;

        NSMutableCharacterSet *endTextMutableCharacterSet = [[NSCharacterSet newlineCharacterSet] mutableCopy];
        [endTextMutableCharacterSet addCharactersInString:@"\""];
        [endTextMutableCharacterSet addCharactersInString:[separator substringToIndex:1]];
        endTextCharacterSet = endTextMutableCharacterSet;
    }
}


- (NSInteger)numberOfNonEmtyFieldNames:(NSArray *)array
{
    NSInteger count = [array count];

    for (NSString *name in array)
        if ([name isEqualToString:@""])
            count --;

    return count;
}


- (NSArray *)parseTable
{
    while (! [scanner isAtEnd]) {
        [self parseEmptyLines];

        NSUInteger location = [scanner scanLocation];

        for (NSString *separatorString in @[ @";", @",", @"\t" ]) {
            [self setSeparator:separatorString];
            [scanner setScanLocation:location];

            fieldNames = [self parseHeader];

            if (fieldNames != nil && [fieldNames count] > 1 && [self parseLineSeparator])
                goto foundHeader;
        }

        [self skipLine];
    }

foundHeader:

    if ([scanner isAtEnd])
        return nil;

    NSMutableArray *records = [NSMutableArray array];

    if ([self numberOfNonEmtyFieldNames:fieldNames] < 2)
        return records;

    NSDictionary *record = [self parseRecord];

    if (!record)
        return records;

    while (record) {
        @autoreleasepool {
            [records addObject:record];

            if (![self parseLineSeparator])
                break;

            record = [self parseRecord];
        }
    }

    return records;
}


- (NSMutableArray *)parseHeader
{
    NSString *name = [self parseField];

    if (!name)
        return nil;

    NSMutableArray *names = [NSMutableArray array];

    while (name) {
        [names addObject:[CSVParser simplifyCSVHeaderName:name]];

        if (![self parseSeparator])
            break;

        name = [self parseField];
    }

    return names;
}


- (NSDictionary *)parseRecord
{
    if ([self parseEmptyLines])
        return nil;

    if ([scanner isAtEnd])
        return nil;

    NSString *field = [self parseField];

    if (!field)
        return nil;

    NSInteger fieldNamesCount = [fieldNames count];
    NSInteger fieldCount = 0;

    NSMutableDictionary *record = [NSMutableDictionary dictionaryWithCapacity:[fieldNames count]];

    while (field) {
        NSString *fieldName;

        if (fieldNamesCount > fieldCount) {
            fieldName = fieldNames[fieldCount];
        } else {
            fieldName = [NSString stringWithFormat:@"FIELD_%ld", (long)(fieldCount + 1)];
            [fieldNames addObject:fieldName];
            fieldNamesCount++;
        }

        record[fieldName] = field;
        fieldCount++;

        if (! [self parseSeparator])
            break;

        field = [self parseField];
    }

    return record;
}


- (NSString *)parseField
{
    [scanner scanCharactersFromSet:[NSCharacterSet whitespaceCharacterSet] intoString:NULL];

    NSString *escapedString = [self parseEscaped];

    if (escapedString)
        return escapedString;

    NSString *nonEscapedString = [self parseNonEscaped];

    if (nonEscapedString)
        return nonEscapedString;

    NSInteger currentLocation = [scanner scanLocation];

    if ([self parseSeparator] || [self parseLineSeparator] || [scanner isAtEnd]) {
        [scanner setScanLocation:currentLocation];
        return @"";
    }

    return nil;
}


- (NSString *)parseEscaped
{
    if (! [self parseDoubleQuote])
        return nil;

    NSString *accumulatedData = [NSString string];

    while (YES) {
        NSString *fragment = [self parseTextData];

        if (!fragment) {
            fragment = [self parseSeparator];

            if (!fragment) {
                fragment = [self parseLineSeparator];

                if (!fragment) {
                    if ([self parseTwoDoubleQuotes])
                        fragment = @"\"";
                    else
                        break;
                }
            }
        }

        accumulatedData = [accumulatedData stringByAppendingString:fragment];
    }

    if (![self parseDoubleQuote])
        return nil;

    return accumulatedData;
}


- (NSString *)parseNonEscaped
{
    return [self parseTextData];
}


- (NSString *)parseTwoDoubleQuotes
{
    if ([scanner scanString:@"\"\"" intoString:NULL])
        return @"\"\"";

    return nil;
}


- (NSString *)parseDoubleQuote
{
    if ([scanner scanString:@"\"" intoString:NULL])
        return @"\"";

    return nil;
}


- (NSString *)parseSeparator
{
    if ([scanner scanString:separator intoString:NULL])
        return separator;

    return nil;
}


- (NSString *)parseEmptyLines
{
    NSString *matchedNewlines = nil;

    NSUInteger location = [scanner scanLocation];

    [scanner scanCharactersFromSet:[NSCharacterSet whitespaceCharacterSet]
                        intoString:&matchedNewlines];

    if (matchedNewlines == nil) {
        [scanner scanCharactersFromSet:[NSCharacterSet characterSetWithCharactersInString:@",;"]
                            intoString:&matchedNewlines];
    }

    if (matchedNewlines == nil)
        matchedNewlines = @"";

    if ([self parseLineSeparator] == nil) {
        [scanner setScanLocation:location];
        return nil;
    }

    return matchedNewlines;
}


- (NSString *)parseLineSeparator
{
    if ([scanner scanString:@"\n" intoString:NULL])
        return @"\n";

    return nil;
}


- (NSString *)skipLine
{
    [scanner scanUpToCharactersFromSet:[NSCharacterSet newlineCharacterSet] intoString:NULL];
    return [self parseLineSeparator];
}


- (NSString *)parseTextData
{
    NSString *data = nil;

    [scanner scanUpToCharactersFromSet:endTextCharacterSet intoString:&data];

    return data;
}

@end

File: Classes/CSVParser.swift
//
//  CSVParser.swift
//  kraftstoff
//
//  Created by Ingmar Stein on 06.05.15.
//
//

import Foundation

final class CSVParser {
	private var csvString: String

	private var separator: String! {
		didSet {
			assert(separator.count == 1)
			endTextCharacterSet = CharacterSet.newlines.union(CharacterSet(charactersIn: "\"" + separator))
		}
	}
	private var scanner: Scanner

	private var fieldNames = [String]()
	private var endTextCharacterSet = CharacterSet()

	static func simplifyCSVHeaderName(_ header: String) -> String {
		return header.replacingOccurrences(of: "[? :_\\-]+",
		                                   with: "",
		                                   options: .regularExpression).uppercased()
	}

	init(inputCSVString: String) {
		// Convert DOS and legacy Mac line endings to Unix
		csvString = inputCSVString.replacingOccurrences(of: "\r\n?", with: "\n", options: .regularExpression)

		scanner = Scanner(string: csvString)
		scanner.charactersToBeSkipped = nil
	}

	func revertToBeginning() {
		scanner.currentIndex = scanner.string.startIndex
	}

	private func numberOfNonEmtyFieldNames(_ array: [String]) -> Int {
		return array.reduce(0) { (count, name) in name.isEmpty ? count : count + 1 }
	}

	func parseTable() -> [CSVRecord]? {
		scannerLoop: while !scanner.isAtEnd {
			parseEmptyLines()

			let location = scanner.currentIndex

			for separatorString in [ ";", ",", "\t" ] {
				separator = separatorString
				scanner.currentIndex = location

				fieldNames = parseHeader() ?? []

				if fieldNames.count > 1 && parseLineSeparator() != nil {
					break scannerLoop
				}
			}

			skipLine()
		}

		if scanner.isAtEnd {
			return nil
		}

		var records = [CSVRecord]()

		if numberOfNonEmtyFieldNames(fieldNames) < 2 {
			return records
		}

		var record = parseRecord()

		if record == nil {
			return records
		}

		while record != nil {
			var keepGoing = true

			autoreleasepool {
				records.append(record!)

				if parseLineSeparator() == nil {
					keepGoing = false
				} else {
					record = parseRecord()
				}
			}

			if !keepGoing {
				break
			}
		}

		return records
	}

	private func parseHeader() -> [String]? {
		var name = parseField()

		if name == nil {
			return nil
		}

		var names = [String]()

		while name != nil {
			names.append(CSVParser.simplifyCSVHeaderName(name!))

			if parseSeparator() == nil {
				break
			}

			name = parseField()
		}

		return names
	}

	private func parseRecord() -> CSVRecord? {
		if parseEmptyLines() != nil {
			return nil
		}

		if scanner.isAtEnd {
			return nil
		}

		var field = parseField()

		if field == nil {
			return nil
		}

		var fieldNamesCount = fieldNames.count
		var fieldCount = 0

		var record = CSVRecord(minimumCapacity: fieldNamesCount)

		while field != nil {
			let fieldName: String

			if fieldNamesCount > fieldCount {
				fieldName = fieldNames[fieldCount]
			} else {
				fieldName = "FIELD_\(fieldCount + 1)"
				fieldNames.append(fieldName)
				fieldNamesCount += 1
			}

			record[fieldName] = field
			fieldCount += 1

			if parseSeparator() == nil {
				break
			}

			field = parseField()
		}

		return record
	}

	private func parseField() -> String? {
		_ = scanner.scanCharacters(from: CharacterSet.whitespaces)

		if let escapedString = parseEscaped() {
			return escapedString
		}

		if let nonEscapedString = parseNonEscaped() {
			return nonEscapedString
		}

		let currentLocation = scanner.currentIndex

		if parseSeparator() != nil || parseLineSeparator() != nil || scanner.isAtEnd {
			scanner.currentIndex = currentLocation
			return ""
		}

		return nil
	}

	private func parseEscaped() -> String? {
		if parseDoubleQuote() == nil {
			return nil
		}

		var accumulatedData = ""

		while true {
			var fragment = parseTextData()

			if fragment == nil {
				fragment = parseSeparator()

				if fragment == nil {
					fragment = parseLineSeparator()

					if fragment == nil {
						if parseTwoDoubleQuotes() != nil {
							fragment = "\""
						} else {
							break
						}
					}
				}
			}

			accumulatedData += fragment!
		}

		if parseDoubleQuote() == nil {
			return nil
		}

		return accumulatedData
	}

	private func parseNonEscaped() -> String? {
		return parseTextData()
	}

	private func parseTwoDoubleQuotes() -> String? {
		return scanner.scanString("\"\"")
	}

	private func parseDoubleQuote() -> String? {
		return scanner.scanString("\"")
	}

	private func parseSeparator() -> String? {
		return scanner.scanString(separator)
	}

	@discardableResult private func parseEmptyLines() -> String? {
		let location = scanner.currentIndex

		var matchedNewlines = scanner.scanCharacters(from: CharacterSet.whitespaces)

		if matchedNewlines == nil {
			matchedNewlines = scanner.scanCharacters(from: CharacterSet(charactersIn: ",;"))
		}

		if matchedNewlines == nil {
			matchedNewlines = ""
		}

		if parseLineSeparator() == nil {
			scanner.currentIndex = location
			return nil
		}

		return matchedNewlines as String?
	}

	private func parseLineSeparator() -> String? {
		return scanner.scanString("\n")
	}

	@discardableResult private func skipLine() -> String? {
		_ = scanner.scanUpToCharacters(from: CharacterSet.newlines)
		return parseLineSeparator()
	}

	private func parseTextData() -> String? {
		return scanner.scanUpToCharacters(from: endTextCharacterSet)
	}

}
